### PR TITLE
Compute dependency routes

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
Have `DependencyGraph` compute route from root file/notebook to child file/notebook. That route will be later used to populate an `InheritedContext`
Have dependencies convey a property for inheriting context, such that context will be computed when but only when required
This is in preparation of https://github.com/databrickslabs/ucx/pull/2236

### Linked issues
Progresses https://github.com/databrickslabs/ucx/issues/2155
Progresses https://github.com/databrickslabs/ucx/issues/2156

### Functionality
None

### Tests
- [x] added unit tests

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/graph.py [hunk 1]
```diff
@@ -199,6 +199,46 @@ def visit(self, visit_node: Callable[[DependencyGraph], bool | None], visited: s
...
+        assert maybe.graph  # assume caller knows what they're doing
+        if not maybe.graph.dependency.inherits_context:
+            return [maybe.graph.dependency]
+        # handle generic case
+        route = self._do_compute_route(root, leaf, visited)
+        return self._trim_route(route)
+
+    def _do_compute_route(self, root: Path, leaf: Path, visited: set[Path]) -> list[Dependency]:
+        maybe = self.locate_dependency(root)
+        assert maybe.graph  # assume caller knows what they're doing
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/graph.py:214]: don't assume things like that. return a dependency problem if this occurs.

File: src/databricks/labs/ucx/source_code/graph.py [hunk 2]
```diff
@@ -199,6 +199,46 @@ def visit(self, visit_node: Callable[[DependencyGraph], bool | None], visited: s
     def new_dependency_graph_context(self):
         return DependencyGraphContext(parent=self, path_lookup=self._path_lookup, session_state=self._session_state)
 
+    def _compute_route(self, root: Path, leaf: Path, visited: set[Path]) -> list[Dependency]:
+        # leaf is not on the route so handle case where it doesn't inherit context
+        maybe = self.locate_dependency(leaf)
+        assert maybe.graph  # assume caller knows what they're doing
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/graph.py:205]: don't assume things like that. explicitly return a dependency problem if this occurs.

File: src/databricks/labs/ucx/source_code/linters/files.py [hunk 3]
```diff
@@ -38,21 +38,21 @@ def __init__(self, path: Path, source: str, language: Language):
         # using CellLanguage so we can reuse the facilities it provides
         self._language = CellLanguage.of_language(language)
 
+    @property
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/files.py:41]: why do we need this?

File: tests/unit/source_code/samples/functional/parent_that_magic_runs_child_that_uses_value_from_parent.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Databricks notebook source
+
+some_table_name = "some_table_name"
```
Comment (nfx) [inline tests/unit/source_code/samples/functional/parent_that_magic_runs_child_that_uses_value_from_parent.py:3]:
```suggestion
some_table_name = "old.things"
```
use one of the tables from mock migration index to trigger the `table-is-migrated` problem

File: tests/unit/source_code/test_functional.py [hunk 5]
```diff
@@ -69,7 +69,7 @@ class Functional:
 
     @classmethod
     def all(cls) -> list[Functional]:
-        return [Functional(path) for path in cls._location.glob('**/*.py')]
+        return [Functional(path) for path in cls._location.glob('**/*.py') if not path.name.startswith("_")]
```
Comment (nfx) [inline tests/unit/source_code/test_functional.py:72]: please document why it's necessary

File: tests/unit/source_code/test_graph.py [hunk 6]
```diff
@@ -38,8 +38,94 @@ def test_folder_loads_content(mock_path_lookup):
...
+class _TestDependencyGraph(DependencyGraph):
+    def compute_route(self, root: Path, leaf: Path) -> list[Dependency]:
+        return self._compute_route(root, leaf, set())
+
+
+def test_graph_computes_magic_run_route(mock_path_lookup, simple_dependency_resolver):
+    parent = mock_path_lookup.cwd / "functional/parent_that_magic_runs_child_that_uses_value_from_parent.py"
+    child = mock_path_lookup.cwd / "functional/_child_that_uses_value_from_parent.py"
+    dependency = Dependency(NotebookLoader(), parent)
+    root_graph = _TestDependencyGraph(
```
Comment (nfx) [inline tests/unit/source_code/test_graph.py:57]: you can make the constructor of _TestDependencyGraph simpler to fit test scenarios - it's already repeated 5 times, for example.

**Review by nfx — CHANGES_REQUESTED** [review]
make error handling more explicit

File: src/databricks/labs/ucx/source_code/graph.py [hunk 7]
```diff
@@ -199,6 +199,46 @@ def visit(self, visit_node: Callable[[DependencyGraph], bool | None], visited: s
...
+        if not maybe.graph.dependency.inherits_context:
+            return [maybe.graph.dependency]
+        # handle generic case
+        route = self._do_compute_route(root, leaf, visited)
+        return self._trim_route(route)
+
+    def _do_compute_route(self, root: Path, leaf: Path, visited: set[Path]) -> list[Dependency]:
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/graph.py:212]: What does this method do? Could you add some docs?

File: src/databricks/labs/ucx/source_code/graph.py [hunk 8]
```diff
@@ -212,14 +252,19 @@ class DependencyGraphContext:
 
 class Dependency(abc.ABC):
 
-    def __init__(self, loader: DependencyLoader, path: Path):
+    def __init__(self, loader: DependencyLoader, path: Path, inherits_context=True):
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/graph.py:255]:
```suggestion
    def __init__(self, loader: DependencyLoader, path: Path, inherits_context: bool = True):
```

File: src/databricks/labs/ucx/source_code/graph.py [hunk 9]
```diff
@@ -353,7 +395,7 @@ def _local_file_resolver(self) -> BaseFileResolver | None:
     def build_notebook_dependency_graph(self, path: Path, session_state: CurrentSessionState) -> MaybeGraph:
         """Builds a dependency graph starting from a notebook. This method is mainly intended for testing purposes.
         In case of problems, the paths in the problems will be relative to the starting path lookup."""
-        maybe = self._notebook_resolver.resolve_notebook(self._path_lookup, path)
+        maybe = self._notebook_resolver.resolve_notebook(self._path_lookup, path, False)
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/graph.py:398]: Why is the `inherit_context` `False` here? And could you make this a keyword argument to improve readability?

Comment (JCZuurmond) [reply src/databricks/labs/ucx/source_code/graph.py:398]: Same question as in `jobs.py`

**Review by JCZuurmond — COMMENTED** [review]
I have some contextual questions:

File: src/databricks/labs/ucx/source_code/graph.py (hunk 1 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:214]: done

File: src/databricks/labs/ucx/source_code/graph.py (hunk 2 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:205]: done

File: src/databricks/labs/ucx/source_code/linters/files.py (hunk 3 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/files.py:41]: It was already there, it was just moved up (properties before methods)

File: tests/unit/source_code/samples/functional/parent_that_magic_runs_child_that_uses_value_from_parent.py (hunk 4 above)
Comment (ericvergnaud) [reply tests/unit/source_code/samples/functional/parent_that_magic_runs_child_that_uses_value_from_parent.py:3]: we're already triggering the `default-format-changed-in-dbr8` problem, why do you want to trigger 2 problems ?

File: tests/unit/source_code/test_functional.py (hunk 5 above)
Comment (ericvergnaud) [reply tests/unit/source_code/test_functional.py:72]: done

File: tests/unit/source_code/test_graph.py (hunk 6 above)
Comment (ericvergnaud) [reply tests/unit/source_code/test_graph.py:57]: done

File: src/databricks/labs/ucx/source_code/graph.py (hunk 7 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:212]: done

File: src/databricks/labs/ucx/source_code/graph.py (hunk 8 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:255]: not sure that's good practice ? Can you provide a reference ?

File: src/databricks/labs/ucx/source_code/graph.py (hunk 9 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:398]: done

File: tests/unit/source_code/samples/functional/parent_that_magic_runs_child_that_uses_value_from_parent.py (hunk 4 above)
Comment (ericvergnaud) [reply tests/unit/source_code/samples/functional/parent_that_magic_runs_child_that_uses_value_from_parent.py]: done

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**